File: lua/problem_67.lua
local M = {}

function M.new_lfu_cache(cap)
    local c = {cap=cap, keys={}, freqs={}, min_freq=0}

    local function touch(key)
        local e = c.keys[key]
        local bucket = c.freqs[e.freq]
        for i, k in ipairs(bucket) do
            if k == key then table.remove(bucket, i); break end
        end
        if #bucket == 0 then
            c.freqs[e.freq] = nil
            if c.min_freq == e.freq then c.min_freq = e.freq + 1 end
        end
        e.freq = e.freq + 1
        if not c.freqs[e.freq] then c.freqs[e.freq] = {} end
        table.insert(c.freqs[e.freq], key)
    end

    function c:get(key)
        if not self.keys[key] then return nil end
        touch(key)
        return self.keys[key].val
    end

    function c:set(key, val)
        if self.cap == 0 then return end
        if self.keys[key] then
            self.keys[key].val = val; touch(key); return
        end
        local size = 0; for _ in pairs(self.keys) do size = size + 1 end
        if size >= self.cap then
            local evict = self.freqs[self.min_freq][1]
            table.remove(self.freqs[self.min_freq], 1)
            self.keys[evict] = nil
        end
        self.keys[key] = {val=val, freq=1}
        if not self.freqs[1] then self.freqs[1] = {} end
        table.insert(self.freqs[1], key)
        self.min_freq = 1
    end

    return c
end

function M.problem67()
    print("Problem 67:")
    local c = M.new_lfu_cache(2)
    c:set(1,1); c:set(2,2)
    assert(c:get(1) == 1)
    c:set(3,3)
    assert(c:get(2) == nil)
    assert(c:get(3) == 3)
    print("All Tests Passed")
end

return M

File: rust/problem_67.rs
use std::collections::HashMap;

pub struct LFUCache67 {
    cap: usize,
    keys: HashMap<i32, (i32, usize)>,
    freqs: HashMap<usize, Vec<i32>>,
    min_freq: usize,
}

impl LFUCache67 {
    pub fn new(cap: usize) -> Self { LFUCache67 { cap, keys: HashMap::new(), freqs: HashMap::new(), min_freq: 0 } }
    fn touch(&mut self, key: i32) {
        let (val, freq) = *self.keys.get(&key).unwrap();
        let bucket = self.freqs.get_mut(&freq).unwrap();
        bucket.retain(|&k| k != key);
        if bucket.is_empty() { self.freqs.remove(&freq); if self.min_freq == freq { self.min_freq += 1; } }
        let new_freq = freq + 1;
        self.freqs.entry(new_freq).or_default().push(key);
        self.keys.insert(key, (val, new_freq));
    }
    pub fn get(&mut self, key: i32) -> Option<i32> {
        if !self.keys.contains_key(&key) { return None; }
        self.touch(key); Some(self.keys[&key].0)
    }
    pub fn set(&mut self, key: i32, val: i32) {
        if self.cap == 0 { return; }
        if self.keys.contains_key(&key) { self.keys.get_mut(&key).unwrap().0 = val; self.touch(key); return; }
        if self.keys.len() >= self.cap {
            let bucket = self.freqs.get_mut(&self.min_freq).unwrap();
            let evict = bucket.remove(0);
            self.keys.remove(&evict);
        }
        self.keys.insert(key, (val, 1));
        self.freqs.entry(1).or_default().push(key);
        self.min_freq = 1;
    }
}

pub fn problem67() {
    println!("Problem 67:");
    let mut c = LFUCache67::new(2);
    c.set(1,1); c.set(2,2);
    assert_eq!(c.get(1), Some(1));
    c.set(3,3);
    assert_eq!(c.get(2), None);
    assert_eq!(c.get(3), Some(3));
    println!("All Tests Passed");
}
